src/check.c

```diff
@@ -995,3 +1000,210 @@ int scan_root(DOS_FS * fs)
+        switch (get_choice(1, "  Auto-removing label from boot sector.",
+                           2,
+                           1, "Remove invalid label from boot sector",
+                           2, "Enter new label")) {
+        case 1:
+            write_boot_label(fs, "NO NAME    ");
+            memcpy(fs->label, "NO NAME    ", 11);
+            return 1;
+        case 2:
+            get_new_label(doslabel);
+            write_boot_label(fs, doslabel);
+            memcpy(fs->label, doslabel, 11);
+            return 1;
+        }
+    } else if (ret & 0x1) {
+        printf("Label '%s' stored in boot sector contains lowercase characters.\n", pretty_label(fs->label));
+        switch (get_choice(1, "  Ignoring.",
+                           2,
+                           1, "Ignore it",
+                           2, "Change lowercase characters to uppercase")) {
+        case 2:
+            memcpy(doslabel, fs->label, 11);
+            for (i = 0; i < 11; ++i) {
+                /* Fix only locale independent 7bit ASCII portion of characters */
+                if (doslabel[i] >= 'a' && doslabel[i] <= 'z')
+                    doslabel[i] = toupper(doslabel[i]);
+            }
+            write_boot_label(fs, doslabel);
+            memcpy(fs->label, doslabel, 11);
+            return 1;
+        }
+    }
+
+    return 0;
+}
+
+void check_label(DOS_FS *fs)
+{
+    DIR_ENT de;
+    off_t offset;
+    char buffer[256];
+    char doslabel[12];
+    int ret;
+    int i;
+
+    offset = find_volume_de(fs, &de);
+
+    if (offset == 0 && memcmp(fs->label, "NO NAME    ", 11) != 0)
+        check_boot_label(fs);
+
+    if (offset == 0 && memcmp(fs->label, "NO NAME    ", 11) != 0) {
+        printf("Label in boot sector is '%s', but there is no volume label in root directory.\n", pretty_label(fs->label));
+        switch (get_choice(1, "  Auto-removing label from boot sector.",
+                           2,
+                           1, "Remove label from boot sector",
+                           2, "Copy label from boot sector to root directory")) {
+        case 1:
+            write_boot_label(fs, "NO NAME    ");
+            memcpy(fs->label, "NO NAME    ", 11);
+            break;
+        case 2:
+            write_volume_label(fs, fs->label);
+            offset = find_volume_de(fs, &de);
+            break;
+        }
+    }
+
+    if (offset != 0) {
+        memcpy(doslabel, de.name, 11);
+        if (doslabel[0] == 0x05)
+            doslabel[0] = 0xe5;
+        ret = validate_volume_label(NULL, (unsigned char *)doslabel);
+        if (ret & ~0x1) {
+            printf("Volume label '%s' stored in root directory is not valid.\n", pretty_label(doslabel));
+            switch (get_choice(1, "  Auto-removing label.",
+                               2,
+                               1, "Remove invalid label",
+                               2, "Set new label")) {
```

Comment:
This also should have an ‘ignore’ option.

I often set up labels of fixed file systems to be the same as their respective mount points for easier identification (e.g. in GRUB `ls -l`, or to have convenient handles in `/dev/disk/by-label`; `e2fsck` even displays the label instead of the device node’s pathname when checking the file system); this includes setting the label of the EFI system partition to `/boot/efi`. While it may be invalid according to the FAT specification, the violation is pretty benign. Even Microsoft’s checking tools don’t consistently detect it: the DOS version of ScanDisk in Windows 98 ignores it, although the GUI version replaces `/` with `-` and converts the label to uppercase. (DOS utilities mostly ignoring volume labels is probably one reason why Microsoft chose the ‘volume label’ attribute bit for LFN entries, after all.) Otherwise most software only reads the label to display it to the user.

Reply:
Hm... Current fatlabel and mkfs.fat code in git does not allow you to set such label (as they are invalid according to specification). But if you are saying that such labels works fine, and more important that there is also usecase for it, I would this about it a bit more... What about adding a new command line option for these tools which cause that tools would not treat such labels as invalid? I do not think it should be a default behavior of dosfstools to create a filesystem which is invalid according to specification, but if there is reasonable argument for such behavior, additional option could allow usage of it.

Reply:
I use version 4.1 (2017-01-24) from Debian repositories, which still allows it. And I wouldn’t mind it hidden behind a command-line switch, although can be hard to pass command-line flags to boot-time fsck. Maybe an environment variable would work better.

Reply:
I think that for full label check & fix support would be better to first have PR https://github.com/dosfstools/dosfstools/pull/133 merged and this PR rebased on top of them. This label check & fix support started to be more complicated.